src/v/cluster/archival/archival_policy.cc

```diff
@@ -358,10 +358,12 @@ ss::future<std::optional<std::error_code>> get_file_range(
         upl->max_timestamp = seek.ts;
     }
     // Recompute content_length based on file offsets
-    vassert(
-      upl->file_offset <= upl->final_file_offset,
-      "Invalid upload candidate {}",
-      upl);
+    if (upl->sources.size() == 1 && upl->file_offset > upl->final_file_offset) {
```

Comment:
Q: why does `upl->sources.size()` need to equal `1` for this race condition?

Reply:
I guess that here it's not necessary to check this because sources.size() is always 1 (this path never merges data from several segments). Potentially, the upload may contain multiple segments. In this case it's totally fine for it to have any start/stop file offsets because the upload begins in one segment and ends in another segment.

Reply:
I can removes this because it's unnecessary and is probably confusing. WDYT?

Reply:
Personally, I would remove if it it's unnecessary/confusing.

The comment below (`// Normally this shouldn't happen...`) might also be better if it were altered to highlight under what circumstances this path _would_ be taken.

Reply:
done, please take a look

Reply:
Thanks! :+1: